nibabel/tests/test_funcs.py

```diff
@@ -30,34 +30,71 @@ def _as_fname(img):
 
 
 def test_concat():
-    shape = (1,2,5)
-    data0 = np.arange(10).reshape(shape)
-    affine = np.eye(4)
-    img0_mem = Nifti1Image(data0, affine)
-    data1 = data0 - 10
-    img1_mem = Nifti1Image(data1, affine)
-    img2_mem = Nifti1Image(data1, affine+1)
-    img3_mem = Nifti1Image(data1.T, affine)
-    all_data = np.concatenate(
-        [data0[:,:,:,np.newaxis],data1[:,:,:,np.newaxis]],3)
-    # Check filenames and in-memory images work
-    with InTemporaryDirectory():
-        imgs = [img0_mem, img1_mem, img2_mem, img3_mem]
-        img_files = [_as_fname(img) for img in imgs]
-        for img0, img1, img2, img3 in (imgs, img_files):
-            all_imgs = concat_images([img0, img1])
-            assert_array_equal(all_imgs.get_data(), all_data)
-            assert_array_equal(all_imgs.affine, affine)
-            # check that not-matching affines raise error
-            assert_raises(ValueError, concat_images, [img0, img2])
-            assert_raises(ValueError, concat_images, [img0, img3])
-            # except if check_affines is False
-            all_imgs = concat_images([img0, img1])
-            assert_array_equal(all_imgs.get_data(), all_data)
-            assert_array_equal(all_imgs.affine, affine)
-        # Delete images as prophylaxis for windows access errors
-        for img in imgs:
-            del(img)
+    for shape in ((1,2,5), (7,3,1), (13,11,11), (0,1,1)):
+        numel = np.asarray(shape).prod()
+        data0 = np.arange(numel).reshape(shape)
+        affine = np.eye(4)
+        img0_mem = Nifti1Image(data0, affine)
+        data1 = data0 - 10
+        img1_mem = Nifti1Image(data1, affine)
+        img2_mem = Nifti1Image(data1, affine+1)
+        img3_mem = Nifti1Image(data1.T, affine)
+        all_data = np.concatenate(
+            [data0[:,:,:,np.newaxis],data1[:,:,:,np.newaxis]],3)
+        # Check filenames and in-memory images work
+        with InTemporaryDirectory():
+            imgs = [img0_mem, img1_mem, img2_mem, img3_mem]
+            img_files = [_as_fname(img) for img in imgs]
+            for img0, img1, img2, img3 in (imgs, img_files):
+                all_imgs = concat_images([img0, img1])
+                assert_array_equal(all_imgs.get_data(), all_data)
+                assert_array_equal(all_imgs.affine, affine)
+                # check that not-matching affines raise error
+                assert_raises(ValueError, concat_images, [img0, img2])
+                assert_raises(ValueError, concat_images, [img0, img3])
+                # except if check_affines is False
+                all_imgs = concat_images([img0, img1])
+                assert_array_equal(all_imgs.get_data(), all_data)
+                assert_array_equal(all_imgs.affine, affine)
+            # Delete images as prophylaxis for windows access errors
+            for img in imgs:
+                del(img)
+
+        # Test axis parameter and trailing unary dimension
```

Comment:
It would be good to check that values also match what you get from doing np.concatenate directly (as well as checking the shape).  Maybe you could do something like:

for axis in range(-3, 4):
    # some test

Maybe also check axis=None explicitly?